### PR TITLE
Emit StorageDead along unwind paths for generators

### PR DESCRIPTION
**Pull request by tmandry**
Completion of the work done in #60840. That PR made a change to implicitly consider a local `StorageDead` after Drop, but that was incorrect for DropAndReplace (see also #61060 which tried to fix this in a different way).

This finally enables the optimization implemented in #60187.

r? @eddyb 
cc @Zoxc @cramertj @RalfJung

### CODE REVIEW

File: src/librustc_mir/build/scope.rs [hunk 1]
```diff
@@ -735,10 +726,10 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
             // Note that this code iterates scopes from the inner-most to the outer-most,
             // invalidating caches of each scope visited. This way bare minimum of the
             // caches gets invalidated. i.e., if a new drop is added into the middle scope, the
-            // cache of outer scope stays intact.
-            scope.invalidate_cache(!needs_drop, this_scope);
+            // cache of outer scpoe stays intact.
```
Comment (RalfJung) [inline src/librustc_mir/build/scope.rs]:
```suggestion
            // cache of outer scope stays intact.
```

**Comment by cramertj** [comment]
@bors try

(looks like the last one was cancelled somehow?)

**Comment by tmandry** [comment]
@rust-timer build 611f90c11abff7436605a0868f5e37efc1dae6a5

**Comment by cramertj** [comment]
Looks like noise.

**Comment by tmandry** [comment]
The test I just added feels slightly brittle, but I documented what can safely change. I'm open to suggestions on improving this.

I think it's good to have a test that shows the effects of this PR, and it might catch things that the optimization integration test won't.

**Review by eddyb — APPROVED** [review]
LGTM, but if possible I'd like @arielb1, @pnkfelix or @nikomatsakis to take a look at this.

**Comment by arielb1** [comment]
Is it possible to do the same merging approach I did in LLVM, where 2 locals are not considered as overlapping if there is no path between their `StorageLive` statements?

**Comment by arielb1** [comment]
see https://reviews.llvm.org/D31583 and https://bugs.llvm.org//show_bug.cgi?id=32488

**Comment by tmandry** [comment]
@arielb1 hmm, I actually briefly thought about doing something like this, but decided it must be unsound in some way.

Given that it works in LLVM, it might work here. I can't think of a reason why not, at least...

**Comment by RalfJung** [comment]
Well, any execution where they are both live must have executed a `StorageLive` for each of them -- thus witnessing a path from one `StorageLive` to another. With no such path, there can be no such execution. Right?

**Comment by tmandry** [comment]
> Well, any execution where they are both live must have executed a `StorageLive` for each of them

@eddyb pointed out to me earlier, this is true as long as the program is not doing any sort of dynamic trickery where it keeps track of what's `StorageLive` at runtime. e.g.

```rust
if pred { StorageLive(a); } else { StorageLive(b); }
// ...later...
if pred { a = 42; } else { b = 99; }
```

I don't think there's a way to actually write this in Rust, and it should almost certainly be UB. But we should be verifying this in MIR. It can be verified by checking that every usage of `a` is dominated by `StorageLive(a)`. (also @eddyb's idea)

Given that the stabilization that depends on this PR is blocking async/await stabilization, I'm inclined to land this as-is, and then revert it and update the analysis in a follow-up change. This PR and #60187 are correct in their current state, albeit suboptimal. I fear that what should be a "quick fix" might stir up more controversy, or end up being less quick than I'm anticipating :)

**Review by eddyb — APPROVED** [review]

**Comment by Centril** [comment]
Failed in https://github.com/rust-lang/rust/pull/61507#issuecomment-498494928, @bors r-

**Comment by RalfJung** [comment]
> pointed out to me earlier, this is true as long as the program is not doing any sort of dynamic trickery where it keeps track of what's StorageLive at runtime. e.g.

I don't follow -- in your example, it is impossible for `a` and ´b` to be live in the same run. So how is that a counterexample to my statement?

**Comment by tmandry** [comment]
> I don't follow -- in your example, it is impossible for `a` and ´b` to be live in the same run. So how is that a counterexample to my statement?

You're right, it's not a counterexample. We probably shouldn't support it, though.

**Comment by tmandry** [comment]
Failure isn’t reproducing after rebase, resubmitting. 

@bors r=eddyb

**Comment by tmandry** [comment]
Ignoring wasm targets in mir-opt test now.

@bors r=eddyb

**Pull request merged**